harness/determined/profiler.py

```diff
@@ -0,0 +1,559 @@
+    the sender_thread and the sys_metric_collector_thread. They will be cleaned up
+    either once end_after_batch is finished or after 5 minutes have passed.
+
+    You can also ship Timings through the ProfilerAgent with the record_timing() method. This
+    functionality has not yet been implemented - we need to batch the timings and periodically
+    flush them to the sender_thread.
+
+    Profiling is only active between start_on_batch and end_after_batch. It will also automatically
+    shut down 5 minutes after starting. When profiling is not active, no system metrics are
+    collected and the record_timing function is a no-op.
+
+    Usage:
+    ```
+    agent = ProfilerAgent(self, trial_id, agent_id, master_url, start_on_batch, end_after_batch)
+
+    for batch_idx, batch in enumerate(batches):
+        agent.update_batch_idx(batch_idx)
+
+        # NOTE: Timing API has not been fully developed yet
+        forward_pass_timing = Timing("forward_pass")
+        forward_pass_timing.start()
+        # Do forward pass
+        forward_pass_timing.end()
+        agent.record_timing(forward_pass_timing)
+    ```
+    """
+
+    def __init__(
+        self,
+        trial_id: int,
+        agent_id: str,
+        master_url: str,
+        start_on_batch: int,
+        end_after_batch: Optional[int] = None,
+    ):
+        self.current_batch_idx = 0
+        self.agent_id = agent_id
+        self.trial_id = trial_id
+        self.master_url = master_url
+        self.start_on_batch = start_on_batch
+        self.end_after_batch = end_after_batch
+
+        # Set up timer to stop collecting after 5 minutes
+        self.has_started = False
+        self.has_finished = False
+        self.max_collection_seconds = 300
+        self.shutdown_timer = threading.Timer(self.max_collection_seconds, self._end_collection)
+        self.shutdown_timer.daemon = True
+        self.shutdown_lock = threading.Lock()
+
+        # Set up the thread responsible for making API calls
+        self.send_queue: """queue.Queue[Union[
+            List['TrialProfilerMetricsBatch'],
+            'ShutdownMessage',
+        ]]""" = queue.Queue()
+        self.sender_thread = ProfilerSenderThread(self.send_queue, self.master_url)
+        self.sender_thread.start()
+
+        # Launch the system metric collecting thread, but not the actual collection of metrics
+        self.sys_metric_collector_thread = SysMetricCollectorThread(
+            trial_id, agent_id, self.send_queue
+        )
+        self.sys_metric_collector_thread.start()
+
+        # TODO: Add data structure to batch timings and then send to SenderThread
+        #       Does this need to be its own thread to flush correctly?
+        # self.timings_batcher = TimingsBatcher()
+
+    @property
+    def is_enabled(self) -> bool:
+        return self.has_started and not self.has_finished
+
+    def update_batch_idx(self, new_batch_idx: int) -> None:
```

Comment:
q: should this be called for each batch? If so, probably add a check here.

Reply:
That's an interesting point. It should be called for every batch, but I'm not certain if it's incorrect if it isn't, particularly when we think about it being used outside of our harness/with the generic APIs. We'll talk about this at standup.

Reply:
It looks like `new_batch_idx` should be at least larger than or equal to `current_batch_idx`.

Reply:
Fixed